### ISSUE TITLE
Whisper performance: mlxrs vs WhisperKit — improvement plan + benchmark harness

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/mlxrs#344]
# Whisper performance: mlxrs vs WhisperKit — improvement plan + benchmark harness

Pixel-level comparison of the mlxrs Whisper port (`src/audio/stt/`) against Argmax **WhisperKit** (Swift/Core ML), component by component, with each proposed improvement adversarially verified. Includes a runnable benchmark harness (added to the repo) and an honest experimental protocol.

> **Scope/caveats.** WhisperKit runs on the **Apple Neural Engine (Core ML)**; mlxrs runs on the **Metal GPU (MLX)** — MLX has no ANE path, so raw wall-clock/energy is **backend-confounded**, not a fixable gap. This analysis is **static** (the authoring env has no Metal compiler and cannot build MLX); the harness below is for running on a Metal Mac, and the code claims must be re-verified on-machine.

## Root cause behind almost every gap: no `mx.compile`

mlxrs has **no graph compilation/caching anywhere** — every encoder forward and every per-token decoder forward **re-traces the whole graph and eagerly dispatches** each op through mlx-c, vs WhisperKit's compile-once Core ML graphs. Sharpening the earlier finding: the **FFI symbols are already generated** (`mlxrs-sys/src/generated/bindings.rs:1561-1597`: `mlx_compile`, `mlx_set_compile_mode`) — what's missing is a safe Rust wrapper + closure trace-capture shim, **not a bindgen pass**.

## Prioritized improvements (each survived adversarial verification)

| # | Improvement | Location | Effort | Honest expected effect |
|---|---|---|---|---|
| 1 | **Word-timestamps: reuse the already-computed encoder output** instead of re-encoding the 30s window for the alignment pass | `timing.rs:356` (`find_alignment`→`forward_with_cross_qk`) re-encodes at `model.rs:794`; the encoder output is already available at `decoding.rs:1332` (currently dropped) | low | Eliminates **one full encoder forward per window when word-timestamps are on**; zero effect when off. `forward_with_cross_qk` already early-returns on encoder-shaped input (`decoding.rs:1244`), so thread the cached `audio_features` through `apply_word_timestamps`→`find_alignment`. Bit-identical. |
| 2 | **In-place KV cache**: pre-allocate a fixed-capacity self-attn buffer and write 1-token slices via `slice_update`, instead of `concatenate` each step | `layers.rs:437-455` (concat at 443-444); cache type `decoder.rs:36` | high | Turns cumulative O(L·n_state·T²) realloc/copy into O(L·n_state·T) + removes the per-step `Vec` rebuild. Does **not** change FLOPs. Modest on short 30s windows, grows with token count. **Prerequisite for #3** (shape stability for compile caching). |
| 3 | **Bind a safe `mx.compile` wrapper and compile the per-token decoder step** | `decoder.rs:208-313` (`TextDecoder::run`); FFI at `bindings.rs:1568` | high | Removes the per-token graph re-trace + per-op eager dispatch — the **single largest structural gap** vs Core ML — and unlocks MLX fusion. Large fraction of per-token CPU overhead on tiny/base; smaller on large-v3 (matmul-bound). **Depends on #2** (the growing concat cache changes input shape every step, defeating the compile cache). Must re-verify bit-parity on a fixed seed. |
| 4 | **`clip_timestamps` sub-ranges**: skip caller-known non-speech spans | `decoding.rs:1554` seek loop + `TranscribeOptions` (`decoding.rs:1421`) | low | Cheapest path to WhisperKit-style silence skipping (mirrors `prepareSeekClips`). Pure API addition; default = one clip `[0, content_frames]`, byte-identical to today. Pairs with #5. |
| 5 | **Pre-encode energy-gate VAD**: skip silent 30s windows before encode+decode | `decoding.rs:1556` (before `decode_with_fallback`) | medium | For true-silence windows, skips a full encoder forward + ≥1 decoder prefill. Medium-large win on sparse-speech audio. **Must default OFF** to preserve openai-whisper parity; threshold must be conservative (mel-log energy is coarse) or it drops quiet speech (WER risk). |
| 6 | **Defer/pipeline per-step host syncs** (streaming mel + greedy decode) | `mel_spectrogram.rs:369-373`; `decoding.rs:637,672` | medium | Removes forced GPU→CPU round-trips that block consecutive chunks/steps from overlapping. Precondition for #3 delivering value. |
| 7 | **Mixed-bit quant config**: keep the WER-sensitive weight-tied embedding/logit head higher-precision | `lm/quant.rs:51-66` per-layer override | low | Approximates OD-MBP's per-layer idea with **existing code only** (pure config). Caveat: decoder input embedding and output logit head are **one weight-tied tensor** (`decoder.rs:53,265,311`) — one override raises precision for both; heuristic, not OD-MBP's calibrated WER search. |
| 8 | Cache the STFT Hann window + pre-transposed mel filterbank | `audio.rs:128,175`; `mel_spectrogram.rs:361` | trivial | **Effectively no visible latency win** (transpose is a free metadata view; window built once/30s). List as cleanup + a de-risking first consumer of the compile binding. |

## Already at parity — do NOT change or re-implement

- **Cross-attn KV reuse** — projected once per segment, reused (`layers.rs:448-453`).
- **Encoder reuse across temperature fallback** — encoded once at `decoding.rs:1332`, threaded into every retry (the "reuse encoder across temps" idea is already done — a no-op).
- **Dequant-free quantized inference** — `quantized_matmul` on packed weights, no dense reconstruction in the decode loop.
- **Encoder weight quantization** — same per-layer `MaybeQuantizedLinear` path as the decoder; whether a checkpoint quantizes the encoder is a packaging choice, not a code gap.
- **Content-frame-bounded seek**, **on-device logit filters** (suppress/timestamp rules), **batched prompt prefill** (one `(1,T)` forward — *ahead* of WhisperKit's token-by-token prompt feed), **offset-aware causal mask slicing**, **cache-friendly host DTW**, **LocalAgreement streaming** (`streaming/types.rs` provisional→confirmed) — all already implemented.

## Don't pursue (verified dead-ends)

- **Block-diagonal encoder attention mask on stock checkpoints** — WER regression; stock Whisper trained with full attention. Only valid on block-trained converted weights, which mlxrs's ecosystem lacks.
- **Restoring batched best-of / `n_group>1`** — breaks the single-utterance invariant threaded through `decoding.rs`; not a perf win for greedy.
- **Per-window encoder truncation of zero-pad** — self-attention is global, not separable; breaks parity.
- **A streaming worker-pool concurrency layer** — MLX dispatches to a single Metal device (no ANE), so "parallelism" only hides CPU dispatch latency for large new surface area.

## Benchmark harness (added to the repo — uncommitted, compile-unverified on this env)

- `mlxrs/benches/whisper_bench.rs` — extended harness: default full-pipeline RTF + `--component-breakdown` (mel ms, encoder ms/window, first-token latency, mean per-token ms, tokens/sec, segment count, temp-fallback lower-bound, peak MLX-GPU + coarse OS RSS). All APIs verified to exist (`AutoregressiveStt::{log_mel,encode,new_cache,decode_step}`, `memory::peak_memory`, `transforms::eval`). *(Already registered in `Cargo.toml:149`.)*
- `bench/whisperkit_bench.sh` — runs `whisperkit-cli`/`argmax-cli` over the same audio, emits `TranscriptionTimings` JSON, columns mapped 1:1 to the mlxrs metrics.
- `docs/benchmarks/whisper-mlxrs-vs-whisperkit-protocol.md` — runbook: matched models (large-v3-turbo), SHA-pinned LibriSpeech corpus (short/medium/long), 2 warmup + 10 timed (median + MAD), and the fair-vs-confounded metric split.

Run on a Metal Mac:
```
cargo bench --bench whisper_bench --features whisper --release -- \
  <model_dir> <audio-16k.wav> --component-breakdown --reps 7 --lang en
bash bench/whisperkit_bench.sh    # same audio, WhisperKit
```

### Comparability rules (must be honored when reporting)
- **Fairly comparable (algorithmic, head-to-head):** WER/CER (same normalizer), output token count, #encoder passes, #decoder steps over generated tokens, #temperature fallbacks, #segments.
- **Backend-confounded (per-stack only, never a "winner"):** raw ms, RTF, tokens/sec, energy, peak memory — dominated by ANE-vs-GPU + compiled-vs-eager + stateful-vs-concat-KV. Report WhisperKit's `cpuAndGPU` arm as the closest silicon point.
- Known harness gaps: mlxrs non-streaming result has no timing struct (RTF via external wall-clock); word-timestamps aren't exposed in `TranscribeOptions` yet (`timing.rs` unwired) → mark mlxrs **N/A** for the DTW arm until wired.

## Suggested order

1, then 7 (low-risk, low-effort, real wins) → 6 → **2 → 3** (the co-dependent structural pair; benchmark on a Metal Mac) → 4/5 (silence skipping). #1 is implementable and testable **without** Metal (deterministic encode reuse + existing decoder tests).

---
*Comparison verified against WhisperKit source (`argmaxinc/WhisperKit` → `argmax-oss-swift`) and its paper (arXiv:2507.10860). Static analysis only — re-verify on a Metal Mac before instrumenting.*
